src/fu-engine.c

```diff
@@ -1473,16 +1473,25 @@ fu_engine_install (FuEngine *self,
 					     "No cabinet archive to schedule");
 			return FALSE;
 		}
-		return fu_plugin_runner_schedule_update (plugin,
-							 device,
-							 blob_cab,
-							 error);
+		return fu_plugin_runner_schedule_update (plugin, device, blob_cab, error);
 	}
 
 	/* install firmware blob */
 	version_orig = g_strdup (fu_device_get_version (device));
-	if (!fu_engine_install_blob (self, device, blob_fw2, flags, error))
+	if (!fu_engine_install_blob (self, device, blob_fw2, flags, &error_local)) {
+		fu_device_set_update_state (device, FWUPD_UPDATE_STATE_FAILED);
+		fu_device_set_update_error (device, error_local->message);
+		if ((flags & FWUPD_INSTALL_FLAG_NO_HISTORY) == 0 &&
+		    !fu_history_modify_device (self->history, device,
+					       FU_HISTORY_FLAGS_MATCH_OLD_VERSION |
+					       FU_HISTORY_FLAGS_MATCH_NEW_VERSION,
+					       error)) {
+			return FALSE;
+		}
+		fu_device_set_status (device, FWUPD_STATUS_IDLE);
```

Comment:
I think you should set status to idle before modifying history database in case history database modification failed.

Reply:
Agreed, thanks.